homeassistant/components/todo/__init__.py

```diff
@@ -185,21 +214,45 @@ async def _async_create_todo_item(entity: TodoListEntity, call: ServiceCall) ->
 async def _async_update_todo_item(entity: TodoListEntity, call: ServiceCall) -> None:
     """Update an item in the To-do list."""
     item = TodoItem.from_dict(call.data)
-    if item.uid:
-        await entity.async_update_todo_item(item=item)
-        return
+    if not item.uid:
+        found = _find_by_summary(item.summary, entity.todo_items)
+        if not found:
+            raise vol.Invalid(
```

Comment:
We're no longer in the service validation, but in the service handler here. Then we should raise `ValueError` for bad user input and `HomeAssistantError` for api call failures and similar errors. `vol.Invalid` should only be raised from within a `vol.Schema` validation.

Reply:
I remember we have an arch discussion to start on this one... Updated!